docs/incubator-proposal.md

```diff
@@ -0,0 +1,188 @@
+applications require X.509 certificates for use with TLS. Right now, users can
+provide certificates to workloads on the cluster by creating `Secret` resources
+that contain a private/public keypair, or potentially generating a CSR within a
+pod and requesting it be signed by a certificate authority.
+
+A number of projects (see: [kube-lego](https://github.com/jetstack/kube-lego),
+[kube-cert-manager](https://github.com/PalmStoneGames/kube-cert-manager)
+and [kubernetes-letsencrypt](https://github.com/tazjin/kubernetes-letsencrypt))
+came about in order to automate the retrieval of certificates from an [ACME](https://github.com/ietf-wg-acme/acme/)
+compliant server. This has served the community well, especially with the rise
+in popularity of ACME due to [letsencrypt](https://letsencrypt.org). These
+projects also attempt to renew the certificate as it reaches its expiry time.
+
+Whilst each project has invented its own schema/methodology for obtaining
+these certificates, the end goal has largely been the same. To provide a signed
+certificate and private key in a Secret resource to then be consumed, either by
+an Ingress controller or some other workload in-cluster, and then keep that
+certificate up to date.
+
+Furthermore, it is not currently possible with any of these projects to support
+a mode of operation where the private key is kept secret at all times (i.e. the
+key never leaves the application container). It'd also be beneficial for users
+to be able to utilise alternative issuer backends, other than ACME. This could
+include [Hashicorp Vault](https://vaultproject.io), a plain CA (backed by
+Kubernetes secret resources), or any other cloud based service.
+
+It is therefore proposed that a new project, `cert-manager`, is created that
+aims to generalise the issuance and renewal of certificates from an arbitrary
+certificate source. This project should be designed in a similar manner to
+other Kubernetes controllers (eg. kube-controller-manager, service-catalog et
+al.). Whether certificates are signed by a publicly trusted authority is not a
+concern of cert-manager.
+
+cert-manager would also like to support this operating mode where the private
+key is never exposed outside of the application requesting a certificate.
+
+## Definitions
+
+* `Issuer` - an Issuer is a generic backend that issues certificates. cert-manager
+will contain logic on how to process each issuer, and multiple Issuer resources
+utilising the same issuer implementation may exist (e.g. to allow issuing
+certificates from both letsencrypt production & staging). An example manifest for
+an Issuer resource:
+
+```yaml
+kind: Issuer
+metadata:
+  name: letsencrypt-staging
+spec:
+  acme:
+    # The ACME server URL
+    server: https://acme-staging.api.letsencrypt.org/directory
+    # Email address used for ACME registration
+    email: user@example.com
+    # Name of a secret used to store the ACME account private key
+    privateKey: letsncrypt-staging
```

Comment:
Probably too detailed discussion for the purpose of this PR, but adding the comment here anyway since I started thinking through the implications and I wondered if you'd considered alternate approaches for key handling.

This makes an implicit security availability tradeoff that is perhaps worth stating explicitly. Rather than storing the private key for use with letsencrypt we could generate a new one each time the controller starts and hold it in memory. This has the benefit of not relying on the security of the secrets system, so that an attacker would likely need to get root on the cert manager node to be able to extract the key from memory and mint certificates outside the system. In the future the key could be held inside a TPM.

However, this is somewhat contrary to the integration advice, which points out that you may hit a rate limit with fairly disastrous consequences:
https://letsencrypt.org/docs/integration-guide/

And since we're putting the certificates inside secrets for availability reasons and they are valid for 90 days you could just steal those. If I read the docs right the private key is only good for 30 days before you need to re-up on the validation, so the utility of exporting the key is also time-bound.